File: android/src/main/java/com/aoneahsan/unifiedtracking/providers/ProviderManager.java
package com.aoneahsan.unifiedtracking.providers;

import android.content.Context;
import android.util.Log;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ProviderManager {
    private static final String TAG = "ProviderManager";
    private final Context context;
    private final List<AnalyticsProvider> analyticsProviders;
    private final List<ErrorTrackingProvider> errorTrackingProviders;
    private boolean debugMode = false;

    public ProviderManager(Context context) {
        this.context = context;
        this.analyticsProviders = new ArrayList<>();
        this.errorTrackingProviders = new ArrayList<>();
    }

    public void addAnalyticsProvider(AnalyticsProvider provider) {
        if (provider == null) {
            return;
        }
        analyticsProviders.add(provider);
        if (debugMode) {
            Log.d(TAG, "Added analytics provider: " + provider.getName());
        }
    }

    public void addErrorTrackingProvider(ErrorTrackingProvider provider) {
        if (provider == null) {
            return;
        }
        errorTrackingProviders.add(provider);
        if (debugMode) {
            Log.d(TAG, "Added error tracking provider: " + provider.getName());
        }
    }

    public void trackEvent(String event, Map<String, Object> properties) {
        for (AnalyticsProvider provider : analyticsProviders) {
            try {
                provider.trackEvent(event, properties);
            } catch (Exception e) {
                Log.e(TAG, "Failed to track event with " + provider.getName(), e);
            }
        }
    }

    public void identifyUser(String userId, Map<String, Object> traits) {
        for (AnalyticsProvider provider : analyticsProviders) {
            try {
                provider.identifyUser(userId, traits);
            } catch (Exception e) {
                Log.e(TAG, "Failed to identify user with " + provider.getName(), e);
            }
        }
    }

    public void setUserProperties(Map<String, Object> properties) {
        for (AnalyticsProvider provider : analyticsProviders) {
            try {
                provider.setUserProperties(properties);
            } catch (Exception e) {
                Log.e(TAG, "Failed to set user properties with " + provider.getName(), e);
            }
        }
    }

    public void logError(Exception error, Map<String, Object> context) {
        for (ErrorTrackingProvider provider : errorTrackingProviders) {
            try {
                provider.logError(error, context);
            } catch (Exception e) {
                Log.e(TAG, "Failed to log error with " + provider.getName(), e);
            }
        }
    }

    public void setUserContext(String userId, Map<String, Object> properties) {
        for (ErrorTrackingProvider provider : errorTrackingProviders) {
            try {
                provider.setUserContext(userId, properties);
            } catch (Exception e) {
                Log.e(TAG, "Failed to set user context with " + provider.getName(), e);
            }
        }
    }

    public void logRevenue(double amount, String currency, String productId, int quantity, Map<String, Object> properties) {
        for (AnalyticsProvider provider : analyticsProviders) {
            try {
                provider.logRevenue(amount, currency, productId, quantity, properties);
            } catch (Exception e) {
                Log.e(TAG, "Failed to log revenue with " + provider.getName(), e);
            }
        }
    }

    public void logScreenView(String screenName, Map<String, Object> properties) {
        for (AnalyticsProvider provider : analyticsProviders) {
            try {
                provider.logScreenView(screenName, properties);
            } catch (Exception e) {
                Log.e(TAG, "Failed to log screen view with " + provider.getName(), e);
            }
        }
    }

    public void setConsent(boolean analytics, boolean errorTracking, boolean personalization) {
        for (AnalyticsProvider provider : analyticsProviders) {
            try {
                provider.setConsent(analytics);
            } catch (Exception e) {
                Log.e(TAG, "Failed to set consent for " + provider.getName(), e);
            }
        }
        
        for (ErrorTrackingProvider provider : errorTrackingProviders) {
            try {
                provider.setConsent(errorTracking);
            } catch (Exception e) {
                Log.e(TAG, "Failed to set consent for " + provider.getName(), e);
            }
        }
    }

    public void reset() {
        for (AnalyticsProvider provider : analyticsProviders) {
            try {
                provider.reset();
            } catch (Exception e) {
                Log.e(TAG, "Failed to reset " + provider.getName(), e);
            }
        }
        
        for (ErrorTrackingProvider provider : errorTrackingProviders) {
            try {
                provider.reset();
            } catch (Exception e) {
                Log.e(TAG, "Failed to reset " + provider.getName(), e);
            }
        }
    }

    /**
     * Returns the names of providers that are both enabled and initialized,
     * grouped by category: { analytics: [names], errorTracking: [names] }.
     */
    public JSObject getActiveProviders() {
        JSObject result = new JSObject();

        JSArray analyticsArray = new JSArray();
        for (AnalyticsProvider provider : analyticsProviders) {
            if (provider != null && provider.isEnabled() && provider.isInitialized()) {
                analyticsArray.put(provider.getName());
            }
        }

        JSArray errorTrackingArray = new JSArray();
        for (ErrorTrackingProvider provider : errorTrackingProviders) {
            if (provider != null && provider.isEnabled() && provider.isInitialized()) {
                errorTrackingArray.put(provider.getName());
            }
        }

        result.put("analytics", analyticsArray);
        result.put("errorTracking", errorTrackingArray);

        return result;
    }

    public void setDebugMode(boolean enabled) {
        this.debugMode = enabled;
        
        for (AnalyticsProvider provider : analyticsProviders) {
            try {
                provider.setDebugMode(enabled);
            } catch (Exception e) {
                Log.e(TAG, "Failed to set debug mode for " + provider.getName(), e);
            }
        }
        
        for (ErrorTrackingProvider provider : errorTrackingProviders) {
            try {
                provider.setDebugMode(enabled);
            } catch (Exception e) {
                Log.e(TAG, "Failed to set debug mode for " + provider.getName(), e);
            }
        }
    }
}

File: ios/Sources/UnifiedTrackingPlugin/ProviderManager.swift
import Foundation

/// Orchestrates the registered analytics + error-tracking providers and fans out
/// every tracking call to the matching set. All calls are defensive: a provider
/// only receives a call when it is both enabled and initialized.
///
/// Thread-safety: calls are kept synchronous and are expected to run on the main
/// thread (Capacitor dispatches plugin calls on the main queue). The underlying
/// SDKs (Firebase Analytics, Sentry) are safe to call from the main thread for
/// the operations used here.
class ProviderManager {
    private var analyticsProviders: [AnalyticsProvider] = []
    private var errorTrackingProviders: [ErrorTrackingProvider] = []
    private var debugMode = false

    // MARK: - Registration

    func addAnalyticsProvider(_ provider: AnalyticsProvider) {
        analyticsProviders.append(provider)
        // Propagate the current debug flag so providers added after enableDebugMode()
        // still honour it.
        provider.setDebugMode(debugMode)
        if debugMode {
            print("[ProviderManager] Added analytics provider: \(provider.name)")
        }
    }

    func addErrorTrackingProvider(_ provider: ErrorTrackingProvider) {
        errorTrackingProviders.append(provider)
        provider.setDebugMode(debugMode)
        if debugMode {
            print("[ProviderManager] Added error tracking provider: \(provider.name)")
        }
    }

    // MARK: - Analytics fan-out

    func trackEvent(_ event: String, properties: [String: Any]) {
        for provider in activeAnalyticsProviders {
            provider.trackEvent(event, properties: properties)
        }
    }

    /// Routes an identify call to analytics providers (`identifyUser`) AND to
    /// error-tracking providers (`setUserContext`) so error reports are attributed
    /// to the same user. The plugin only exposes `identify`, so this is the single
    /// entry point that keeps both sides in sync.
    func identifyUser(_ userId: String, traits: [String: Any]) {
        for provider in activeAnalyticsProviders {
            provider.identifyUser(userId, traits: traits)
        }
        for provider in activeErrorTrackingProviders {
            provider.setUserContext(userId, properties: traits)
        }
    }

    func setUserProperties(_ properties: [String: Any]) {
        for provider in activeAnalyticsProviders {
            provider.setUserProperties(properties)
        }
    }

    func logRevenue(amount: Double, currency: String, productId: String?, quantity: Int, properties: [String: Any]) {
        for provider in activeAnalyticsProviders {
            provider.logRevenue(amount: amount, currency: currency, productId: productId,
                                 quantity: quantity, properties: properties)
        }
    }

    func logScreenView(_ screenName: String, properties: [String: Any]) {
        for provider in activeAnalyticsProviders {
            provider.logScreenView(screenName, properties: properties)
        }
    }

    // MARK: - Error tracking fan-out

    func logError(_ error: Error, context: [String: Any]) {
        for provider in activeErrorTrackingProviders {
            provider.logError(error, context: context)
        }
    }

    // MARK: - Consent / lifecycle (apply to ALL registered providers, regardless of
    // current enabled/initialized state — consent must be able to flip a provider on
    // or off, and reset must run even when disabled).

    /// Routes analytics consent to analytics providers and error-tracking consent to
    /// error providers. `personalization` is accepted for parity with the JS API but
    /// has no distinct native channel today; analytics providers gate personalization
    /// internally via their own consent flag.
    func setConsent(analytics: Bool, errorTracking: Bool, personalization: Bool) {
        for provider in analyticsProviders {
            provider.setConsent(analytics)
        }
        for provider in errorTrackingProviders {
            provider.setConsent(errorTracking)
        }
    }

    func reset() {
        for provider in analyticsProviders {
            provider.reset()
        }
        for provider in errorTrackingProviders {
            provider.reset()
        }
    }

    func setDebugMode(_ enabled: Bool) {
        debugMode = enabled
        for provider in analyticsProviders {
            provider.setDebugMode(enabled)
        }
        for provider in errorTrackingProviders {
            provider.setDebugMode(enabled)
        }
    }

    // MARK: - Introspection

    /// Returns only the providers that are currently enabled AND initialized, as
    /// flat name arrays: `{ "analytics": [names...], "errorTracking": [names...] }`.
    func getActiveProviders() -> [String: Any] {
        return [
            "analytics": activeAnalyticsProviders.map { $0.name },
            "errorTracking": activeErrorTrackingProviders.map { $0.name }
        ]
    }

    // MARK: - Private helpers

    private var activeAnalyticsProviders: [AnalyticsProvider] {
        analyticsProviders.filter { $0.isEnabled && $0.isInitialized }
    }

    private var activeErrorTrackingProviders: [ErrorTrackingProvider] {
        errorTrackingProviders.filter { $0.isEnabled && $0.isInitialized }
    }
}
